src/services/pcn-bridge/src/Bridge.cpp

```diff
@@ -1,149 +1,344 @@
+  std::vector<uint8_t>::const_iterator last = packet.begin() + start + bpdusize;
+  std::vector<uint8_t> p(first, last);
+  getStp(vlan)->processBPDU(port_id, p);
+}
+
+void Bridge::broadcastPacket(Ports &port,
+                             polycube::service::PacketInMetadata &md,
+                             const std::vector<uint8_t> &packet) {
+  logger()->debug("Broadcasting packet..");
+
+  bool tagged = md.metadata[1];
+  uint16_t vlan = md.metadata[0];
+  logger()->trace("metadata vlan:{0} tagged:{1}", vlan, tagged);
+
+  for (auto &it : get_ports()) {
+    if (it->name() == port.name()) {
+      continue;
+    }
+    logger()->debug("Sending packet to port {0}", it->name());
+    it->send_packet_out(packet, tagged, vlan);
+  }
 }
 
-// Basic default implementation, place your extension here (if needed)
 std::shared_ptr<Ports> Bridge::getPorts(const std::string &name) {
-  // call default implementation in base class
   return BridgeBase::getPorts(name);
 }
 
-// Basic default implementation, place your extension here (if needed)
 std::vector<std::shared_ptr<Ports>> Bridge::getPortsList() {
-  // call default implementation in base class
   return BridgeBase::getPortsList();
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::addPorts(const std::string &name, const PortsJsonObject &conf) {
   BridgeBase::addPorts(name, conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::addPortsList(const std::vector<PortsJsonObject> &conf) {
-  // call default implementation in base class
   BridgeBase::addPortsList(conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
-void Bridge::replacePorts(const std::string &name, const PortsJsonObject &conf) {
-  // call default implementation in base class
+void Bridge::replacePorts(const std::string &name,
+                          const PortsJsonObject &conf) {
   BridgeBase::replacePorts(name, conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::delPorts(const std::string &name) {
-  // call default implementation in base class
   BridgeBase::delPorts(name);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::delPortsList() {
-  // call default implementation in base class
   BridgeBase::delPortsList();
 }
 
 std::shared_ptr<Fdb> Bridge::getFdb() {
-  throw std::runtime_error("Bridge::getFdb: Method not implemented");
+  return fdb_;
 }
 
 void Bridge::addFdb(const FdbJsonObject &value) {
-  throw std::runtime_error("Bridge::addFdb: Method not implemented");
+  if (fdb_ != nullptr)
+    throw std::runtime_error("Fdb already exists");
+
+  fdb_ = std::make_shared<Fdb>(*this, value);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::replaceFdb(const FdbJsonObject &conf) {
-  // call default implementation in base class
-  BridgeBase::replaceFdb(conf);
+  delFdb();
```

Comment:
I don't think so, since addFdb is called in the constructor of Bridge. Anyway, in delFdb there is a check on the value of fdb_, so there is no risk.
Should i add a check in replaceFdb too?

Reply:
This is good, actually I don't remember what I was thinking when I asked that question, I don't see any potential problems.